### ISSUE TITLE
Proposal to supplement `--trim` option

### ISSUE THREAD

**Issue by jamesaoverton** [ontodev/robot#483]
`remove` and `filter` include a `--trim <true/false>` option to determine which axioms are removed/filtered. The idea is to trim (or not) any "dangling" references.

In the implementation we build a set of target OWLObjects, then for each axiom test whether the target set intersects ([`getPartialAxioms`](https://github.com/ontodev/robot/blob/master/robot-core/src/main/java/org/obolibrary/robot/RelatedObjectsHelper.java#L147)) or contains ([`getCompleteAxioms`](https://github.com/ontodev/robot/blob/master/robot-core/src/main/java/org/obolibrary/robot/RelatedObjectsHelper.java#L92)) the set of OWLObjects for the axiom. OWLAPI has an [`OWLAxiom.getSignature`](http://owlcs.github.io/owlapi/apidocs_5/org/semanticweb/owlapi/model/HasSignature.html#getSignature--) method that returns the set of (named) OWLEntities for the axiom. But we also wanted to be able to target unnamed OWLObjects, so we wrote a [`getObjects`](https://github.com/ontodev/robot/blob/master/robot-core/src/main/java/org/obolibrary/robot/OntologyHelper.java#L338) method that starts with the `getSignature` set and adds OWLObjects based on the axiom type.

#473 shows that there are cases where we don't care about the unnamed OWLObjects, and `getSignature` would be better than `getObject`.

While maintaining backwards compatibility, we could add an option that lets you use `getSignature` instead of `getObject` and thus only consider the named OWLEntities when checking each axiom. I'm not sure what to call it: `--compare named`, `--compare entities`, `--signature named`.

A distinct but related question is whether we care about axiom annotations when checking each axiom. The current implementation of `getCompleteAxioms` will add the annotated axiom if the annotations are contained in the target set, or the unannotated axiom of only the axiom itself is contained in the target set. This suggests that we could have a more general option that would select from several comparison strategies.

**Comment by dosumis** [comment]
It seems overly complex to add yet another option  &  I suspect `--signature named` will be obscure to anyone not familiar with the details of the OWL-API.  Would it not be reasonable to have 

`--strict true --select anonymous` do this?  It's what I expected that option to do in the first place.  It doesn't seem to do anything much on it's own right now.

**Comment by jamesaoverton** [comment]
There is no `--strict` option. `--trim` is behaving as designed and documented, even if that behaviour is not what  you want, so we won't break that behaviour. The `remove` and `filter` commands are low-level and in this case we expect user to know some details of OWLAPI.

We can add this option now, and we can "bundle"/"alias" the release artifacts into something more friendly once the behaviour is nailed down.

**Comment by cmungall** [comment]
I think @dosumis proposal is to use `--strict true` to indicate the signature is used, but using language that is less OWLAPI centric.

I apologize for not paying attention to this before. I've read through the tickets a few more times and am still having problems mentally reasoning about the different combinations.

I may be misunderstanding but it seems some of the misunderstanding arose from this line in the current docs on http://robot.obolibrary.org/filter

> By default, filter will copy axioms from the input ontology for which all the entities are in the specified set.

I think some of this read this as OWLEntities, or named entities. IMHO it seems unintuitive to include anonymous class expressions in this list, but maybe I misunderstand the use cases for this.

However, arguments can be made that "entity" is sufficiently flexible that either interpretation works.

Could there be an argument that we interpret the current docs to mean named objects, and that therefore it is OK to change the current behavior? I think we should think carefully about this, we'll likely have the set of options we choose now for at least a decade, let's also think about making this easy for users to reason about.

In the other ticket there is talk of an `anonymous-parents` option as well. I worry this has the potential to complicate further. 

Boarding flight now, sorry if this adds more confusion than help...

**Comment by jamesaoverton** [comment]
Part of my frustration here is that `remove` and `filter` were developed in public from August 22, 2017 (#183) to August 4, 2018 (#290) and were released in ROBOT 1.2.0. I have use cases for selecting anonymous objects. I rely on these commands in several projects. I don't want to break code that's using these commands -- my code or anybody else's code. It has been hard to document these commands well, and I'm happy to revise the documentation.

These two commands are specifically designed for expert users who understand OWLAPI. They're supposed to be building blocks for higher-level functionality. Once all the details of the release artifacts are nailed down, we should bundle them into a `robot release` command or something that's very convenient for users. So for these two commands I prefer to use the OWLAPI terminology, even if it's more obscure.

Still, if calling this `--strict true` rather than `--signature true` sounds better to you guys, I guess I'm fine with that name.

I'd still like to hear feedback on my paragraph about axiom annotations. In light of that, I think it would be even better to deprecate `--trim` and use `--compare [partial|named|complete]`, leaving a possibility for other comparison options in the future.

In any case, I'll be much more confident in any solution if we have known-good examples to test.

**Comment by jamesaoverton** [comment]
I think this is better documentation. It would be better still with a concrete example and visuals. The final bit assumes that we use `--compare [partial|named|complete]` rather than `--strict [true|false]`. 

---

`remove` operates in a series of four steps. The first step is to specify a set of terms using `--term` and `--term-file`. These are named `OWLEntity`s, and they constitute the initial "target set" for `remove`. If no terms are specified, then the initial target set includes all the terms in the ontology.

The second step is optional: you can modify the target set using "selectors". Each selector will return a new set of related `OWLObject`s. Each object may be either named or anonymous. Think of the selection step in terms of Venn diagrams for sets of objects in the ontology. If no `--select` option is provided, or you specify `--select self`, then the target set is just the initial target set of terms. If you `--select ancestors` then the new target set is the set of all the ancestors for all the objects in the old target set. Note that this new target set does not include the objects in the old target set. If you `--select "self ancestors"` then the new target set will include *both* the old target set (`self`) and all the ancestors, i.e. the union of all the selections. You can use a series of `--select`s, and each one will operate on the previous target set. So you can specify `--select parents --select children` to get first the parents and then all of their children, effectively selecting your original target terms and all of their siblings.

The third step is to choose which axioms to match and remove. By default, all axioms that match are removed, but you can limit this to any combination of `logical` axioms, `annotation` axioms, `equivalent` axioms, etc.

The fourth step is to match and remove the axioms. By default, if the target set contains one or more objects used in an axiom then the axiom is removed. This is called `--compare partial`. You can be more strict and use `--compare complete`, so that if the target set contains all the objects used in the axiom then the axiom is removed. Keep in mind that this option requires all named *and* anonymous objects (`OWLObject`s) to be in the target set for an axiom to be matched and removed. Your ontology will likely still include "dangling" axioms that use objects from the target set. You can specify `--compare named`, so that if the target set contains all of the named objects (`OWLEntity`s) used in the axiom then the axiom is removed.

For filter the fourth step is opposite, because matched axioms are *copied* to the output ontology. The default is `--compare complete`, so if the target set contains all objects (named and anonymous) used in the axiom then the axiom is copied. You can use `--compare named`, so if the target set contains all named objects used in the axiom then the axiom is copied. Or you can use `--compare partial`, so that if the target set contains one or more objects used in the axiom then the axiom is copied. This will likely copy "dangling" axioms that use objects that are not in the target set.

**Comment by matentzn** [comment]
Ok cool. I think this documentation cleared up some of my fundamental problems with understanding, thanks! 

However, before we zone in on a solution, I would like to clarify one last time. Am I assuming correctly, that, with ROBOT in its current form, it is NOT at all possible to (positively) filter ax2 from the following ontology using only select:

```
ax1: A sub B
ax2: A sub R some C
```

Can you explain why 

```
robot filter --term A --term R --term C --trim true --select "self anonymous" 
```

does NOT select axiom 2 (just one last time, sorry). Then I will give you my position :) Thanks.

**Comment by jamesaoverton** [comment]
Sure, this needs to be more clear. The `parents` selector would work in this small example, but not in the more general case required for #473.

Here's how this example works, following the four steps above:

1. Your `--term` options give you an initial target set of {A, R, C}.

2. The `self` selector takes the initial target set and just returns it: {A, R, C}. the `anonymous` selector takes the initial target set and returns the subset of anonymous `OWLObject`s from that set, which is {}. Then we take the union of those two selector results and get {A, R, C}.

3. By default, all axioms are considered for matches.

4. By default the matching strategy is "complete" (aka `filter --trim true`) and uses [`getObjects`](https://github.com/ontodev/robot/blob/master/robot-core/src/main/java/org/obolibrary/robot/OntologyHelper.java#L338). So the set of objects used in ax2 is {A, R, C, R some C}. The 'R some C' is an anonymous `OWLObject`, and since it is not in the target set, ax2 is not matched.

I have use cases where I want to be able to select anonymous objects, which is why we wrote `getObjects`. You might think that the `anonymous` selector would give you all the anonymous `OWLObject`s in the ontology, or maybe all anonymous objects for {R, C}, but it only gives you the subset of anonymous objects from the target set. This isn't clear in the current docs and could be a main point of confusion.

In this example the `parents` selector would give you {B, R some C}. In #473 the `parents` selector also adds unwanted named objects, so it doesn't solve the problem. In that issue I proposed an `anonymous-parents` selector, which I think would solve the immediate problem. But now I favour a more general solution in which we provide the option to use [`getSignature`](https://owlcs.github.io/owlapi/apidocs_4/org/semanticweb/owlapi/model/HasSignature.html#getSignature--) instead of `getObjects`. Using `getSignature` the set of objects used in ax2 is {A, R, C} (only named objects), all of which are in the target set, so the axiom is matched. The proposals here are various ways provide an option to use `getSignature` instead of `getObjects`.

**Comment by matentzn** [comment]
Okay thanks, now I see where my confusion lies. In your previous documentation, you say:

> Think of the selection step in terms of Venn diagrams for sets of objects in the ontology. If no --select option is provided, or you specify --select self, then the target set is just the initial target set of terms. If you --select ancestors then the new target set is the set of all the ancestors for all the objects in the old target set. Note that this new target set does not include the objects in the old target set. If you --select "self ancestors" then the new target set will include both the old target set (self) and all the ancestors, i.e. the union of all the selections

So in other words: we have an initial set T, and then **add some new stuff** (ancestors!) to that set T. 

In your explanation of "self anonymous" you say:

> the anonymous selector takes the initial target set and returns the subset of anonymous OWLObjects from that set, which is {}. Then we take the union of those two selector results and get {A, R, C}.

Which reads as: "the select statement produces a **subset** of T" (i.e. nothing is added). I guess this is where the confusion lies for me: "self ancestors" expands the set. "self anonymous" reduces it. :P Sorry to drag this out. Its an important feature. And if I have clarity on this issue, I can tell you what I think should happen regarding the additional trim options! Thanks for your patience :P

**Comment by dosumis** [comment]
Hi James.  Many thanks for the detailed explanation.  I now understand your reasoning, but I'm still concerned that we need to be careful to ensure ROBOT behaves intuitively and that the documentation is easy to understand.  We have 7 regular ROBOT users in our group - all of them OWL literate, many with OWL-API experience.  They are *all* using the filter commands and have all experienced serious difficulties in understanding how they works and in using them to produce the outputs they need.  I worry that any fix for this problem that adds yet another switch risks making the command more confusing.

> You might think that the anonymous selector would give you ...  all anonymous objects for {R, C}

That was my assumption (which was shared by others in the office).  I'd like to understand how changing default behaviour to this would break backwards compatibility.

>  but it only gives you the subset of anonymous objects from the target set

Does that mean that "self anonymous" and "self" are equivalent, or is there any circumstance (any target list) where "self anonymous" would return more?  If so, can you give an example?

If not, then I'd like to understand how a change in behaviour to have  "self anonymous" return to all anonymous objects with entities in the target list would cause backwards compatibility problems.  Presumably your concern is for the way this interacts with other `--select` options?  If so, could you give some examples?

**Possible solutions:**

If I understand your proposal correctly, you propose a new option `--compare` with possible values 'partial', 'complete' and 'named'.   

The name: `--compare` sounds opaque and unintuitive to me (what is being compared to what?) and I worry that adding yet another option will make understanding behaviour even harder, e.g. how will `--compare` options interact with `--trim true/false` ?

If backwards compatibility issues resulting from changing the default behaviour of the `--select anonymous` option are too serious, could we instead add a similarly named `--select` value that does give this behaviour?  Not entirely sure what we could name it `--select anonymous_all`  ???

**Comment by dosumis** [comment]
(Note - hadn't seen Nico's post when I added mine)

**Comment by matentzn** [comment]
after we have clarity on "anonymous", I will share my opinion on how to expand. I agree, compare is opaque, and probably a new selector like David says is fine as well (class-expression; complex; expression), even though they would probably all be kind of fake synonyms of "anonymous".

**Comment by jamesaoverton** [comment]
Thanks for the feedback. I think we're making good progress.

@dosumis asked for an example where `anonymous` does some work. Continuing with @matentzn's example, `--select parents` results in {B, R some C}, so `--select parents --select anonymous` would result in {R some C}.

@matentzn contrasts "add some new stuff" vs "subset" behaviour. In the [docs](http://robot.obolibrary.org/remove) there's a list of "Relation Types" selectors such as `ancestors` that use a "relation" to give you a new target set ("new stuff"), and there's a list of "Entity Types" selectors such as `classes` that give you a subset of your target set. `anonymous` might sound like the former, but it's implemented as the latter.

I'm learning a lot from this discussion and I'll revise the documentation.

My impression is that @dosumis finds `getSignature` behaviour more intuitive, and I think that's a more general solution that will serve us better in the long run than a new selector.

My current preference is to deprecate `--trim` in favour of `--compare`, so:

- `filter --trim true` becomes `filter --compare complete` (default)
- `filter --trim false` becomes `filter --compare partial`
- `remove --trim true` becomes `remove --compare partial` (default)
- `remove --trim false` becomes `remove --compare complete`

Then `--compare named` would be like "complete" but use `getSignature`, fixing #473, and giving the option to ignore anonymous objects and focus on named entities.

But if just adding `--strict true` (meaning use `getSignature`) is really what you want, then I'm ok with that. PR #484 implements `--signature true|false` which we could rename to `--strict`, or whatever.

Both @dosumis and @matentzn are suggesting new selectors as an alternative. We're free to add more selectors -- that won't break backward compatibility. Selectors are implemented in [`RelatedObjectsHelper.java`](https://github.com/ontodev/robot/blob/master/robot-core/src/main/java/org/obolibrary/robot/RelatedObjectsHelper.java). Keep in mind that we're relying on `EntitySearcher` to do the heavy lifting.

What should the new selector(s) do? It needs to "add some new stuff". I would implement my suggestion for `anonymous-parents` by calling [`selectParents()`](https://github.com/ontodev/robot/blob/master/robot-core/src/main/java/org/obolibrary/robot/RelatedObjectsHelper.java#L612) then keeping only anonymous objects. If the suggestions for `anonymous_all` or `class-expressions` mean something other than that, please tell me. If you can clarify that for me, I can make a PR to try it out.

@dosumis also asked about changing the behaviour of `anonymous`. If you mean changing `anonymous` to behave like this `anonymous-parents` then that would break backwards compatibility. Using @matentzn's example `filter --term A --select anonymous` gives us {} but `filter --term A --select anonymous-parents` gives us {R some C}. If that's not what you mean, then I need clarification.


No matter what, I would really like known-good example files that we can test to be sure that we're really solving the problem.

**Comment by matentzn** [comment]
Given this example:

```
ax1: A sub B
ax2: A sub R some C
```

What is the result of this:
```
filter --term A --select "self anonymous parents"  --signature true --trim true
```

**Comment by jamesaoverton** [comment]
@matentzn: In #484 the code stays the same until step four, where we decide which axioms to act on:

1. the initial term set is {A}
2. select `self` {A}, `anonymous` {}, `parents` {B, R some C}: so the target set is now the union {A, B, R some C}
3. consider all axioms (default)
4. is ax2 copied? `--signature true` means that the set of named objects for ax2 is {A, R, C}
    A. `filter --trim false` means use [`getPartialAxioms`](https://github.com/ontodev/robot/blob/9083e418ec1eabcf0bbc4f7bdad98241dd48e2e9/robot-core/src/main/java/org/obolibrary/robot/RelatedObjectsHelper.java#L202); so since at least one element of {A, R, C} is in the target set, then ax2 is copied
    B. `filter --trim true` means use [`getCompleteAxioms`](https://github.com/ontodev/robot/blob/9083e418ec1eabcf0bbc4f7bdad98241dd48e2e9/robot-core/src/main/java/org/obolibrary/robot/RelatedObjectsHelper.java#L120); since not every element in {A, R, C} is in the target set, then ax2 is not copied

Note: My plan was to just use `getSignature` but I had to implement [`getNamedObjects`](https://github.com/ontodev/robot/blob/9083e418ec1eabcf0bbc4f7bdad98241dd48e2e9/robot-core/src/main/java/org/obolibrary/robot/OntologyHelper.java#L339) which also includes the subject, predicate, and value of annotations.

This is a good example to consider. I can't tell if the behaviour for (B) is what you want or not, and now I can't tell if it fully addresses #473. This example suggests the need for a `components` (`signatures`?) selector that takes {A, B, R some C} and returns {R, C}. Then `--select "self parents" --select "named components"` would result in {A, B, R, C}.

**Comment by jamesaoverton** [comment]
I updated the documentation for #484. As far as I can tell, it does what we agreed to on Monday. But I still don't have good tests against known-good examples, and @matentzn's last example makes me worry that we still do not have this exactly right.

**Comment by matentzn** [comment]
I think the documented behaviour in step 4 sounds fine to me. And I would hold off on the named_entity selector for now until a proper use case arises - for now, I wanted clarity more than anything, so thankse :).  

> My plan was to just use getSignature but I had to implement getNamedObjects which also includes the subject, predicate, and value of annotations.

So in this example:

```
ax1.1: [A sub B](xref:http://someuri#D)
```
Would

```
filter --term A --term B --signature true --trim true
```

ax1.1. or not? I believe it should; but maybe I misunderstand your use of getNamedObjects().

**Comment by dosumis** [comment]
This discussion of examples is confusing to me as they don't match the motivating use case.  Given this ontology 

```yaml
ax1: A sub B
ax2: A sub R some C
```

+ a term list {A, R, C} - we want to return ax2 but not ax 1

With `--signature true`, what is the minimal `robot filter` that would return this?

Would it be:

```sh
filter --term A --term R --term C  --select self --signature true --trim true
```

term list is {A, R, C} so any axiom with only A, R and C in the signature is return.  

Do we even need --select self here, or is that default behavior?

**Comment by matentzn** [comment]
This is what I understood from the latest documentation! 

```
filter --term A --term R --term C --signature true --trim true
```
Will return ax2 and not ax1.

Now the new question is, given:
```
ax1: A sub B
ax2: A sub R some C [dbxref: B]
```

Given the above query, will ax2 be returned or not?

**Comment by jamesaoverton** [comment]
@matentzn is right that `--select self` is the default, and that `filter --term A --term R --term C --signature true --trim true` will copy ax2 but not ax1.

The xref annotation is a little trickier. With `filter --trim true` we're using `getCompleteAxioms()` which has [special code](https://github.com/ontodev/robot/blob/524b5c7f611a8041450014119dd8f94a44837998/robot-core/src/main/java/org/obolibrary/robot/RelatedObjectsHelper.java#L148) for `OWLAnnotationAssertionAxiom`s:

1. With target set {A, R, C} the underlying axiom `A sub R some C` will be copied, but its xref annotation does not match the target set and will not be copied.
2. With target set {A, R, C, dbxref, B} both ax2 and its xref annotation will be copied.
3. For `filter` there's a special [`annotations` selector](https://github.com/ontodev/robot/blob/524b5c7f611a8041450014119dd8f94a44837998/robot-command/src/main/java/org/obolibrary/robot/FilterCommand.java#L233) that will got back through and add annotations for all the copied objects, so with target set {A, R, C} and `annotations` you should get ax2 and its xref annotation.

In other words, axiom annotations won't prevent the underlying axiom from being matched, but the annotations are only copied if they're in the target set. You need to use the `annotations` selector for the common case where you want to copy all the annotations for the target set, without specifying every annotation property. `annotations` is for `filter` only, not `remove`, and there's a section on it in the docs.

One other weird thing is that `OWLAnnotationAssertionAxioms` work in terms of subject/predicate/value IRIs, not objects, so the special code works with IRIs. I think it's doing the right thing, but I admit that it's confusing.

**Comment by dosumis** [comment]
> In other words, axiom annotations won't prevent the underlying axiom from being matched, but the annotations are only copied if they're in the target set. You need to use the annotations selector for the common case where you want to copy all the annotations for the target set, without specifying every annotation property. 

That sounds like the most intuitive behavior to me.

> One other weird thing is that OWLAnnotationAssertionAxioms work in terms of subject/predicate/value IRIs, not objects, so the special code works with IRIs. I think it's doing the right thing, but I admit that it's confusing.

Can you givea an example of the implications of this?

**Comment by jamesaoverton** [comment]
@matentzn gave two xref examples today: `xref:http://someuri#D` and `dbxref: B`. If `B` is an `OWLEntity` then you can specify it using `--term` or the normal selectors, such as `parents`. But if `http://someuri/#D` is just an IRI, then it's an `OWLObject` but not an `OWLEntity`. The `EntitySearcher` and `ontology.getEntitiesInSignature()` won't see it. So we use `annotation.getValue()` and other code to try and handle plain IRIs transparently, but it's extra complexity. To handle IRI annotations specifically we added the `annotations` selector and the pattern selectors: `--select "xref=http://someuri#D"`.

So we can handle objects and IRIs, but sometimes they require different approaches.

I wish it were simpler. OWL is complicated. OWLAPI is complicated. I tried to design something intuitive: specify your terms, select related terms, remove/filter the axioms with those terms. But at every turn there are more complications: anonymous objects, annotations on axioms, annotations use IRIs not objects, ontology annotations are their own thing, etc.

**Comment by cmungall** [comment]
I need to be able to extract a subset of uberon, starting at nephron, and walking up the partonomy and is-a hierarchy. I am willing to write the documentation for how to do this, but I'm having trouble figuring out if it's even possible with the current implementation.

I'm trying various combinations like this:

```
robot filter  --input uberon.obo -t UBERON:0001285 -t BFO:0000050 -p true --signature true --select "self ancestors annotations anonymous"  --output test.obo
```

but it won't walk the relational graph. It looks like `ancestors` means only subClassOf between named classes, I think many users will assume ancestors means ancestors in the relational graph.

I sympathize with James point in the previous comment. The OWLAPI is hard, and crucially it does not always map to users' conceptual model of an ontology. 

Perhaps the simplest thing here would be to specify a new operation `subgraph`? Filter will remain incredibly useful for certain kinds of use cases (axiom by axiom), but we don't try and further overload it?

**Comment by jamesaoverton** [comment]
Sorry, I'm confused.

As a general concept "ancestors" only makes sense in a tree, not a graph. I can see how we could define a new selector for remove/filter that would climb both the subclass and partonomy trees. I don't know what to call it, but "subgraph" would not be a good name.

Maybe you're proposing a new `SubgraphOperation` class, to handle relations in general (with cycles, reflexivity, symmetry, and transitivity)?

**Comment by matentzn** [comment]
I think he means that if you have a --[part of]--> b that b is an ancestor of a (in the relational graph).

**Comment by jamesaoverton** [comment]
Sure. Your partonomy should form a tree, so "ancestors" can make sense. But that's not true of relations and graphs in general, so I'm thrown by the phrase "relational graph".

**Comment by matentzn** [comment]
Yeah we really should have an agreed upon taxonomy of graph types.. I believe the word relational graph is kind of the sum of all isa, equivalence axioms and existential restrictions - mostly. But yeah, I share your u certainty about the meaning of these terms. I also heard existential graph often

**Comment by jamesaoverton** [comment]
The main issue here was resolved by #484. I'm going to close this issue now. Feel free to refer back to this in new issues.
